File: Logic/Snippets/Java/ValueOrRef.java
import java.util.HashMap;

// 値型と参照型のメモ
public class ValueOrRef {

    public static void main(String args[]) {
        
        Test test = new Test();

        // 値型
        String hoge = "hoge";
        String hoge2 = test.method1(hoge);
        System.out.println(hoge);
        System.out.println(hoge2);

        // 参照型

        String[] strAry = new String[]{"Tokyo", "Osaka", "Hakata"};
        String[] strAry2 = test.method2(strAry);
        System.out.println(strAry[1]);
        System.out.println(strAry2[1]);

        HashMap<String, String> hashmap = new HashMap<String, String>();
        hashmap.put("a", "Tokyo");
        hashmap.put("b", "Osaka");
        hashmap.put("c", "Kobe");
        HashMap<String, String> hashmap2 = test.method3(hashmap);
        System.out.println(hashmap.get("b"));
        System.out.println(hashmap2.get("b"));

        Person person = new Person();
        test.method4(person);
        System.out.println("person.name=" + person.name);

        Person person3 = person; // コピーされずに同じものの参照となる
        person3.name = "Person3";
        System.out.println("person.name=" + person.name); // こちらも"Person3"となる
        System.out.println("person3.name=" + person3.name);
    }
}

class Person {
    String name;
}

class Test {
    // 値型
    
    // 値型の値渡し（String）
    String method1(String str) {
        //str = "XXX" //<- コンパイルエラー
        String ans = str; // 値型のためcopyされて別物となる
        ans = "XXX";
        return ans;
    }
    
    // 参照型

    // 参照型の値渡し（Array）
    String[] method2(String[] array) {
        array[1] = "XXXX"; //<- コンパイルエラーにならない
        String[] ans = array; // 参照
        ans[1] = "ZZZZ";
        return ans;
    }

    // 参照型の値渡し（Dictionary）
    HashMap<String, String> method3(HashMap<String, String> dict) {
        dict.replace("b", "XXXX");
        HashMap<String, String> ans = dict; // 参照
        ans.replace("b", "ZZZZ");
        return ans;
    }

    // 参照型の値渡し
    void method4(Person person) {
        person.name = "Yamada";
    }
    
}

File: Logic/Others/Snippets/Kotlin/ValueOrRef.kt

// 値型と参照型のメモ

class Person(var name: String = "", var age: Int = 0) {
}
data class Robot(var name: String = "", var age: Int = 0) {
}

class Test() {
    
    // 値型
    
    // 値型の値渡し（String）
    fun method7(str: String): String {
        //str = "XXX" //<- コンパイルエラー
        var ans = str // 値型のためcopyされて別物となる
        ans = "XXX"
        return ans
    }
    
    // 参照型
    
    // 参照型の値渡し（Array）
    fun method3(array: Array<String>): Array<String> {
        array[1] = "XXXX" //<- コンパイルエラーにならない
        var ans = array // 参照
        ans[1] = "ZZZZ"
        return ans
    }

    // 参照型の値渡し（Dictionary）
    fun method5(dict: MutableMap<String, String>): MutableMap<String, String> {
        dict["b"] = "XXXX" //<- コンパイルエラーにならない
        var ans: MutableMap<String, String> = dict // 参照
        ans["b"] = "ZZZZ"
        return ans
    }

    // 参照型の値渡し
    fun method10(person: Person) {
        person.name = "Yamada"
    }

    // Data Class
    fun method11(robot: Robot): Robot {
        robot.name = "Yamada" //<- コンパイルエラーにならない
        var ans = robot // 参照
        ans.name = "Takeda"
        return ans
    }
}

fun main() {
    val test = Test()

    // 値型

    var myStr = "Yamaha"
    val myStr2 = test.method7(myStr)
    println("myStr=${myStr}") // 引数で渡したものは変わっていない
    println("myStr2=${myStr2}")

    // 参照型

    var myArray = arrayOf("AAA", "BBB", "CCC")
    val myArray2 = test.method3(myArray)
    println("myArray[1]=${myArray[1]}") // myArray2と同じになる
    println("myArray2[1]=${myArray2[1]}")

    var myDict = mutableMapOf<String, String>("a" to "AAA", "b" to "BBB", "c" to "CCC")
    val myDict2 = test.method5(myDict)
    println("myDict[b]=${myDict["b"]}") // myDict2と同じになる
    println("myDict2[b]=${myDict2["b"]}")

    val person = Person()
    test.method10(person)
    println("person.name=${person.name}")

    val person3 = person // コピーされずに同じものの参照となる
    person3.name = "Person3"
    println("person.name=${person.name}") // こちらも"Person3"となる
    println("person3.name=${person3.name}")

    var robot = Robot()
    robot.name = "Honda"
    var robot2 = test.method11(robot)
    println("robot.name=${robot.name}") // 引数で渡したものも変わっている
    println("robot2.name=${robot2.name}")

    var robot4 = robot2 // 参照
    robot4.name = "Kawasaki"
    println("robot2.name=${robot2.name}") // robot4と同じになる
    println("robot4.name=${robot4.name}")
}

